tm2/pkg/sdk/baseapp.go

```diff
@@ -838,13 +839,93 @@ func (app *BaseApp) runTx(mode RunTxMode, txBytes []byte, tx Tx) (result Result)
 
 // EndBlock implements the ABCI interface.
 func (app *BaseApp) EndBlock(req abci.RequestEndBlock) (res abci.ResponseEndBlock) {
+	// update the minGasPrices in the block header
+	newHeader := app.updateHeader()
 	if app.endBlocker != nil {
 		res = app.endBlocker(app.deliverState.ctx, req)
 	}
-
+	// send updated Header back to consensus
+	res.Header = newHeader
 	return
 }
 
+// it updates the MinGasPrice in BlockHeader of the current block
+// the MiinGasPrice will be used to check transaction for the next block
+func (app *BaseApp) updateHeader() abci.Header {
+	header := app.deliverState.ctx.BlockHeader().(*bft.Header)
+	lastPrice := header.GasPriceAmount
+	gasUsed := app.deliverState.ctx.BlockGasMeter().GasConsumed()
+	// calculate and update the dynamic gas price based on the gas used in the current block
+	header.GasPriceAmount = app.calBlockMinPriceAmount(gasUsed, lastPrice)
+	app.deliverState.ctx = app.deliverState.ctx.WithBlockHeader(header)
+	return header
+}
+
+// it calculates the minGasPrice for the txs to be included in the next block.
+// minGasPrice = lastPrice + lastPrice*(gasUsed-TargetBlockGas)/TargetBlockGas/GasCompressor)
+func (app *BaseApp) calBlockMinPriceAmount(gasUsed int64, lastPrice int64) int64 {
+	// if no last price set
+	if lastPrice == 0 {
+		return 0
+	}
+	// if no gas used, no need to change the lastPrice
+	if gasUsed == 0 {
+		return lastPrice
+	}
+
+	targetGas := app.consensusParams.Block.TargetGas
+	// if used gas is right on target, no need to change
+	if gasUsed == targetGas {
+		return lastPrice
+	}
+
+	var (
+		num   = new(big.Int)
+		denom = new(big.Int)
+	)
+
+	c := app.consensusParams.Block.PriceChangeCompressor
+
+	lastMinPrice := big.NewInt(lastPrice)
+	if gasUsed > targetGas { // increase gas price
```

Comment:
Could this calculation be done in a way that makes it a bit easier to follow based on the formula definition? Something like this?
```go
// Declare big integers for the calculation.
minGasPriceBig := big.NewInt(0) // result value
lastPriceBig := big.NewInt(lastPrice)
gasUsedBig := big.NewInt(gasUsed)
targetGasBig := big.NewInt(targetGas)
gasCompressorBig := big.NewInt(app.consensusParams.Block.PriceChangeCompressor)

// Formula:
// minGasPrice = lastPrice + lastPrice*(gasUsed-TargetBlockGas)/TargetBlockGas/GasCompressor)
minGasPriceBig.Sub(gasUsedBig, targetGasBig)
minGasPriceBig.Mul(minGasPriceBig, lastPriceBig)
minGasPriceBig.Div(minGasPriceBig, targetGasBig)
minGasPriceBig.Div(minGasPriceBig, gasCompressorBig)
minGasPriceBig.Add(minGasPriceBig, lastPriceBig)
```

Reply:
The math formula is just an abstraction of a simple solution for the underlying problem we're trying to solve.

1) What do we do if the gas used is less than the target gas in a block?
2) How do we bring the gas used back to the target level?

We simplify the solution with a one-line formula to explain the idea. However, in reality, we need to treat two scenarios differently. For example, in the first case, we need to increase the gas by at least 1 unit, and in the second case, we should set a floor to the target gas price.

This is just a starting point. Down the line,  the solution might not be even representable by one simple formula